pkg/crc/machine/client.go

```diff
@@ -30,21 +30,30 @@ type Client interface {
 }
 
 type client struct {
-	name   string
-	debug  bool
-	config crcConfig.Storage
+	name           string
+	debug          bool
+	config         crcConfig.Storage
+	virtualMachine VirtualMachine
 
 	diskDetails *memoize.Memoizer
 	ramDetails  *memoize.Memoizer
 }
 
 func NewClient(name string, debug bool, config crcConfig.Storage) Client {
+	return newClientWithVirtualMachine(name, debug, config, nil)
+}
+
+// newClientWithVirtualMachine creates a Client instance with a overridden VirtualMachine implementation.
+// It would not create a new VirtualMachine object. This method is primarily created for usage in tests so
+// that we can pass a fake VirtualMachine implementation.
+func newClientWithVirtualMachine(name string, debug bool, config crcConfig.Storage, vm VirtualMachine) Client {
 	return &client{
-		name:        name,
-		debug:       debug,
-		config:      config,
-		diskDetails: memoize.NewMemoizer(time.Minute, 5*time.Minute),
-		ramDetails:  memoize.NewMemoizer(30*time.Second, 2*time.Minute),
+		name:           name,
+		debug:          debug,
+		config:         config,
+		diskDetails:    memoize.NewMemoizer(time.Minute, 5*time.Minute),
+		ramDetails:     memoize.NewMemoizer(30*time.Second, 2*time.Minute),
+		virtualMachine: vm,
```

Comment:
I have split this PR into multiple commits, please review if it's as per your expectations. I can adjust if additional changes are needed.

Reply:
Hmm https://github.com/cfergeau/crc/commits/pr/issue4407/ is closer to what I had in mind. One commit doing only the VirtualMachine interface introduction and making use of its methods (could be 2 commits if you prefer, but I think it's fine as one). I added 2 follow-up commits which are meant to be squashed in the first one.
Then one commit adds `newClientWithVirtualMachine`, `loadVirtualMachineLazily` so that we can override it, and the rest of your series, I haven't touched it.

Reply:
By the way, the content in `Solutions/Idea` and `Proposed Changes` would be nice to have in the individual commit logs, as one line commit logs are very terse when you'll get back to them in 6 months trying to remember what the commit was about, but also too long for a short log, they overflow quite a bit in github UI, git log, ... They typically should be about 60 characters in length.

Better to have a short 1line log, followed by more details in the log explaining why you are making the change, and giving a high level overview of what you are doing if it's a big change.